wgpu-hal/src/vulkan/drm.rs

```diff
@@ -0,0 +1,146 @@
+        fd: i32,
+        plane: u32,
+        connector_id: u32,
+        width: u32,
+        height: u32,
+        refresh_rate: u32,
+    ) -> Result<super::Surface, crate::InstanceError> {
+        if !self.shared.extensions.contains(&khr::display::NAME) {
+            return Err(crate::InstanceError::new(
+                "Vulkan driver does not support VK_KHR_display".to_string(),
+            ));
+        }
+
+        let drm_stat = {
+            let mut stat = MaybeUninit::<libc::stat>::uninit();
+
+            if unsafe { libc::fstat(fd, stat.as_mut_ptr()) } != 0 {
+                return Err(crate::InstanceError::new(
+                    "Unable to fstat drm device".to_string(),
+                ));
+            }
+
+            unsafe { stat.assume_init() }
+        };
+
+        let raw_devices = match unsafe { self.shared.raw.enumerate_physical_devices() } {
+            Ok(devices) => devices,
+            Err(err) => {
+                log::error!("enumerate_adapters: {}", err);
+                Vec::new()
+            }
+        };
+
+        let mut physical_device = None;
+
+        for device in raw_devices {
+            let properties2 = vk::PhysicalDeviceProperties2KHR::default();
+
+            let mut drm_props = vk::PhysicalDeviceDrmPropertiesEXT::default();
+            let mut properties2 = properties2.push_next(&mut drm_props);
+
+            unsafe {
+                self.shared
+                    .raw
+                    .get_physical_device_properties2(device, &mut properties2)
+            };
+
+            /*
+                The makedev call is just bit manipulation to combine major and minor device numbers into a Unix device ID.
+                It doesn't perform any filesystem operations, only bitshifting.
+                See: https://github.com/rust-lang/libc/blob/268e1b3810ac07ed637d9005bc1a54e49218c958/src/unix/linux_like/linux/mod.rs#L6049
+                We use the resulting device IDs to check if the Vulkan raw device from enumerate_physical_devices
+                matches the DRM device referred to by our file descriptor.
+            */
+
+            let primary_devid =
+                libc::makedev(drm_props.primary_major as _, drm_props.primary_minor as _);
+            let render_devid =
+                libc::makedev(drm_props.render_major as _, drm_props.render_minor as _);
+
+            if primary_devid == drm_stat.st_rdev || render_devid == drm_stat.st_rdev {
+                physical_device = Some(device)
+            }
+        }
+
+        let physical_device = physical_device.ok_or(crate::InstanceError::new(
+            "Failed to find suitable drm device".to_string(),
+        ))?;
+
+        // FIXME: consider implementing this strategy on working vulkan drivers
+        // let displays = unsafe {
+        //     display_instance
+        //         .get_physical_device_display_properties(physical_device)
+        //         .expect("Failed to get displays")
+        // };
```

Comment:
This should either be an issue or be removed.

Reply:
Mostly a left over while I was testing things. I will open a proper tracking issue in the future